test/Icicle/Test/Arbitrary/Source.hs

```diff
@@ -276,3 +182,283 @@ qwfDesugar qwf
+      -- and any nesting will be simplified away
+      num <- oneof_sized [return 0] [return 1]
+      (ctxs,tgi') <- genCtxs num toptgi []
+      x <- genExp tgi'
+      return (Query (reverse ctxs) x)
+
+ where
+  genCtxs :: Int -> TypedGenInfo -> [Context () T.Variable] -> Gen ([Context () T.Variable], TypedGenInfo)
+  genCtxs 0 tgi ss = return (ss, tgi)
+  genCtxs n tgi ss
+   = do (ctx',tgi') <- genCtx tgi
+        genCtxs (n-1) tgi' (ctx' : ss)
+
+  genCtx :: TypedGenInfo -> Gen (Context () T.Variable, TypedGenInfo)
+  genCtx tgi
+   | TTAgg <- tgiTemp tgi
+   = genCtxAgg tgi
+   | otherwise
+   = genLet tgi
+
+  genCtxAgg :: TypedGenInfo -> Gen (Context () T.Variable, TypedGenInfo)
+  genCtxAgg tgi
+   = oneof_sized
+      [ genWindow tgi
+      , nobinds tgi . Latest   () . abs <$> arbitrary ]
+      [ nobinds tgi . GroupBy  () <$> genExp tgi
+      , nobinds tgi . Distinct () <$> genExp tgi
+      , nobinds tgi . Filter   () <$> genExp tgi
+      , genFold tgi
+      , genLet tgi
+      , genGroupFold tgi
+      ]
+
+  genWindow tgi
+   = do pre <- arbitrary
+        post <- arbitrary
+        return (Windowed () pre post, tgi)
+
+  genFold tgi
+   = do v <- arbitrary :: Gen (CB.Name T.Variable)
+        f <- Fold v <$> genExp tgi { tgiTemp = TTPure } <*> genExp tgi { tgiTemp = TTElt, tgiVars = Map.insert v TTElt $ tgiVars tgi } <*> arbitrary
+        let tgi' = tgi { tgiVars = Map.insert v TTAgg $ tgiVars tgi }
+        return (LetFold () f, tgi')
+  genLet tgi
+   = do v <- arbitrary :: Gen (CB.Name T.Variable)
+        tt' <- genTemporality tgi
+        l <- Let () v <$> genExp tgi { tgiTemp = tt' }
+        let tgi' = tgi { tgiVars = Map.insert v tt' $ tgiVars tgi }
+        return (l,tgi')
+  genGroupFold tgi
+   = do k <- arbitrary :: Gen (CB.Name T.Variable)
+        v <- arbitrary :: Gen (CB.Name T.Variable)
+        let tgi' = tgi { tgiVars = Map.insert k TTElt $ Map.insert v TTElt $ tgiVars tgi }
+
+        (Query ctxs x) <- genQuery tgi
+        grpky <- genExp tgi
+        let grp' = Query (GroupBy () grpky : ctxs) x
+
+        let g = GroupFold () k v (Nested () grp')
+        return (g, tgi')
+
+  nobinds tgi x
+   = (x, tgi)
+
+genTemporality :: TypedGenInfo -> Gen TestTemporality
+genTemporality tgi
+ = case tgiTemp tgi of
+    TTPure -> return TTPure
+    TTElt  -> elements [TTPure, TTElt]
+    TTAgg  -> elements [TTPure, TTElt, TTAgg]
+
+
+
+-- Boring arbitrary instances ---
+-- These are all quite simple, and just use the above functions when they aren't
+
+instance Arbitrary T.Variable where
+ arbitrary
+  = T.Variable <$> elements muppets
+
+instance Arbitrary (Exp () T.Variable) where
+ arbitrary
+  = do tt <- arbitrary
+       genExp $ TypedGenInfo Map.empty tt tgiDefaultTableflip
+
+instance Arbitrary Prim where
+ arbitrary
+  = oneof
+        [ Lit . LitInt    <$> pos
+        , Lit . LitDouble <$> pos'
+        , Lit . LitString <$> (elements southpark)
+        , Lit . LitTime   <$> arbitrary ]
```

Comment:
could we generate other prims (such as builtin functions) and generate the arguments for them? It would be nice to have the prims covered by quickcheck